Live m-mapping of chunks on disk

codesome opened the pull request:
This is a broken down piece of https://github.com/prometheus/prometheus/pull/6679/

This just has the standalone implementation of the live m-mapping logic without any integration with Head.

codesome commented on tsdb/docs/format/head_chunks.md:
```diff
@@ -0,0 +1,41 @@
+├──────────────────────────────┤
+│    version(1) <1 byte>       │
+├──────────────────────────────┤
+│    padding(0) <3 byte>       │
```
I am reusing the chunks package to create a segment. So it adds the padding by default.

bwplotka replied:
padding is needed to make sure we have aligned pages

krasi-georgiev replied:
nice, learned something new :)

codesome commented on tsdb/docs/format/head_chunks.md:
````diff
@@ -0,0 +1,41 @@
+
+```
+┌─────────────────────┬───────────────────────┬───────────────────────┬───────────────────┬───────────────┬──────────────┐
+| series ref <8 byte> | mint <8 byte, uint64> | maxt <8 byte, uint64> | encoding <1 byte> | len <uvarint> | data <bytes> |
````
Good point. I will add a crc.

codesome commented on tsdb/chunks/head_chunks_test.go:
```diff
@@ -0,0 +1,285 @@
+
+	mint := binary.BigEndian.Uint64(actualBytes[HeaderMintOffset : HeaderMintOffset+8])
+	maxt := binary.BigEndian.Uint64(actualBytes[HeaderMaxtOffset : HeaderMaxtOffset+8])
+	testutil.Assert(t, mint > 0, "expected mint >0")
```
The chunk data is only to verify that the exact bytes were written and read. We don't need the mint and maxt passed to be of the actual chunk in this case, just that input and output should match.

codesome commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,711 @@
+
+// IterateAllChunks iterates on all the chunks in its byte slices in the order of the segment file sequence
+// and runs the provided function on each chunk. It returns on the first error encountered.
+func (w *HeadReadWriter) IterateAllChunks(f func(seriesRef, chunkRef uint64, mint, maxt int64) error) error {
```
It is used in https://github.com/prometheus/prometheus/pull/6679 to replay the mmapped chunks

codesome commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,711 @@
+	)
+
+	w.wbufLock.RLock()
+	// TODO(codesome): take care of possible race with w.wbuf.
```
That comment is no longer valid, will remove it.

codesome commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,711 @@
+	MagicHeadChunks = 0x0130BC91
+
+	headChunksFormatV1 = 1
+	inBufferShards     = 16
```
I just chose it randomly. Not sure what would be the right size here.

codesome replied:
If there are too many incoming queries, 16 _might_ hurt a bit. I will make it 128 for now and we can modify it later if there are any issues.

codesome replied:
(It's purely a guess, Prombench can help us later)

codesome commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,711 @@
+	inBufferChunksLocks [inBufferShards]sync.RWMutex
+
+	// Reader.
+	bs map[int]ByteSlice
```
I just followed the names in the chunk Reader of the block. I can change the name to be `byteSlice` or something (as they are bytes and don't directly hold chunks)

krasi-georgiev replied:
the signature clearly indicates that these are `ByteSlice`, but the bytes are actual chunks and this makes the code that uses this field easyer to understand.

`chks map[int]ByteSlice`

codesome replied:
Will change

codesome commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,711 @@
+	//
+	// The upper 4 bytes are for the segment index and
+	// the lower 4 bytes are for the segment offset where to start reading this chunk.
+	chunkRef = w.chunkRef(uint64(w.seq()), uint64(w.n))
```
Looks like this is the only place where w.seq() is used. I will remove that method unless required.

codesome commented on tsdb/chunks/head_chunks_test.go:
```diff
@@ -0,0 +1,285 @@
+	"github.com/prometheus/prometheus/util/testutil"
+)
+
+func TestHeadReadWriter_WriteChunk(t *testing.T) {
```
Though they might look similar in some parts, they are functionally testing different things in a slightly different way. Combining them is possible, but it would get messy/confusing real quick. I would like to keep them separate for simplicity.

bwplotka replied:
I am also a fan of table tests, but if you think there is no way to make them generic - fine to keep it separated :+1:

codesome commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,711 @@
+// HeadReadWriter is for writing the Head block chunks to the disk
+// and access chunks via an mmapped file.
+type HeadReadWriter struct {
+	// Writer.
```
While using this in Head, it works like a single component. My reasons for having it like this is: makes it simpler to handle any race that can occur while cutting new segment (where you have to update the mmapped file, etc). Also having it as separate components will add additional complexity in Head as it has to take care of keeping the writer and reader in sync. Head is complex enough, so having it as single component will avoid us from handling it in Head.

bwplotka replied:
I totally see argument for both, but happy with this being single struct.

krasi-georgiev replied:
sno I meant something like
```
HeadR struct{
}
HeadW struct{
}
type HeadReadWriter struct {
    HeadR{ // all reader fields here}
    HeadW{ // all writer fields here}


}
```

codesome replied:
Will check if that leads to a cleaner/simpler implementation.

codesome replied:
The read and write operations need to share a lock. So, at max, we will have the same methods for the original structs while the Read and Write structs will just have the methods which do the basic operation - something similar to just splitting out the methods with meaningful names. 

I will go ahead with having a single struct, as splitting the methods between 2 structs is just an added complexity for the original struct. I will try to structure the code to have the read and write parts separated as much as possible.

codesome commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,711 @@
+		// The file should not be closed if there is no error,
+		// its kept open in the HeadReadWriter.
+		if returnErr != nil {
+			f.Close()
```
We need the new file to be open as it will be the one being appended to. So we don't close inside `cutSegmentFile`. (closing and reopening is additional latency in the hot path)

krasi-georgiev replied:
yep, my bad

codesome commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,711 @@
+		return err
+	}
+
+	n, f, seq, err := cutSegmentFile(w.dirFile, MagicHeadChunks, headChunksFormatV1, 0)
```
Creative way to comment :)

0 so that the file is not preallocated. Current logic around replaying chunks from file depends on the file size, i.e. iterate through the entire file till there are bytes. Preallocation messes it up if the entire file is not filled.

codesome replied:
Any ideas on how I can get rid of that?

bwplotka replied:
it's fine - it's easily visible if you go to `cutSegmentFile`. Good enough for me.

codesome commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,711 @@
+	for i := 0; i < inBufferShards; i++ {
+		w.inBufferChunks[i] = make(map[uint64]chunkenc.Chunk)
+	}
+	for i := inBufferShards - 1; i >= 0; i-- {
```
> a comment why you need to leave one unlocked.

Did not get that. All are locked and unlocked.

And from your comment I realized that these 3 loops can be combined into one :)

bwplotka replied:
Yes (: But iterations looks nice -> index in slice starts from 0 and ends on len()-1 @krasi-georgiev  (:

codesome commented on tsdb/chunks/head_chunks_test.go:
```diff
@@ -0,0 +1,285 @@
+	app, err := chunk.Appender()
+	testutil.Ok(t, err)
+	for i := 0; i < len; i++ {
+		app.Append(rand.Int63(), rand.Float64())
```
It doesn't matter for tests. I am using these chunks only to check the contents. The mint and maxt of the chunks are passed explicitly in the methods inside the tests.

codesome commented on tsdb/chunks/head_chunks_test.go:
```diff
@@ -0,0 +1,285 @@
+
+	expectedBytes := []byte{}
+	numChunks := 2000
+	nextChunkOffset := uint64(HeadSegmentHeaderSize + 24)
```
chunk offset is the one after seriesRef, mint and maxt (hence 24=8+8+8) (the point where the encoding, data len, and the data starts). I have replaced it with constant variables now (yet to push)

codesome commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,711 @@
+		n:           0,
+		segmentTime: segmentTime,
+		pool:        pool,
+		quit:        make(chan struct{}),
```
Boolean also works. Just a habit to use a channel.

codesome replied:
And usually in this case I believe it should be atomic load and store else it would be termed as a race in some cases. I don't see boolean load and store in `atomic` module. Any suggestions?

krasi-georgiev replied:
I don't see where this can cause a race, this api is never to be used concurently right? like the rest of the TSDB api.

codesome replied:
`WriteChunk` and `Chunk` can be called concurrently, it is safe. But I see the closing part is protected by mutex. I will change it to a boolean.

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,711 @@
+
+// Segment header fields constants.
+const (
+	// MagicHeadChunks is 4 bytes at the head of a segment file.
```
segment file? How we name things here? Let's decide and be consistent (: 

I thought segment was `chunk/000001` file? Can we define clear naming?

codesome replied:
I have seen "segment file" being used both for chunk files and the WAL files. So it felt natural to use the same here.

bwplotka replied:
Hm. I think that's creating just confusion. Let's change it then? (:

codesome replied:
Open for suggestions :)

bwplotka replied:
I think we need to start some quick discussion on -developers for that.

> and the WAL files. 

Hm... so there is 3rd segment?

I would propose:

* `index`: **Index File**
* `chunks/000001`: **Chunks File / Segment **
* `wal/chunks/000001`: **Head Chunks File / Head Segment**

But not sure about this 3rd segment you mentioned, what exactly you mean?

codesome replied:
> But not sure about this 3rd segment you mentioned, what exactly you mean?

https://github.com/prometheus/prometheus/blob/46129fdebbbc1f8c73b450b3629b92bfad6b2d95/tsdb/wal/wal.go#L75-L80
WAL in itself consists of multiple files where each file is called a segment :P

bwplotka replied:
:cry:

bwplotka replied:
ok so

I would propose:

* `index`: **Index File**
* `chunks/000001`: **Chunks File / Segment**
* `wal/chunks/000001`: **Head Chunks File / WAL Chunks Segment**
* https://github.com/prometheus/prometheus/blob/46129fdebbbc1f8c73b450b3629b92bfad6b2d95/tsdb/wal/wal.go#L75-L80: **WAL Segment**

codesome replied:
I have no strong opinions on that. I will use "head chunk file" then and replace wherever I find the word "segment"

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,711 @@
+)
+
+var (
+	// DefaultHeadChunkSegmentTime is the default chunks segment time range.
```
time range?

Can we be explicit here? (: I think it's the segmentMaxTime kind of? as it's a threshold when we start to create segments, no?

codesome replied:
Yeah. I will modify the name.

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,711 @@
+	return hrw, nil
+}
+
+func (w *HeadReadWriter) initReader() (err error) {
```
If we can, it would be awesome to be explicit what we do. `init` is not really helpful. What about `openMemMapFiles?`

codesome replied:
Will update the function names

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,711 @@
+		return err
+	}
+	if w.pool == nil {
+		w.pool = chunkenc.NewPool()
```
Did we benchmark if this pool is really helping?

codesome replied:
This is not happening right now, which is the same case with getting chunks from the blocks where we take from the pool and never put back :). I will be passing the pool present in the DB when integrating this. I would prefer keeping it like that so that if we get around putting it back to pool safely, it would be easy integration with this too (automatically, as we are using the same pool).

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,711 @@
+
+// Repair deletes all the segments after the one which had the corruption
+// (including the corrupt file).
+func (w *HeadReadWriter) Repair(originalErr error) error {
```
Hm it looks like it deletes stuff not repairs (:

codesome replied:
That is how current WAL Repair works too, it deletes everything after the corruption point.

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,711 @@
+
+// HeadReadWriter is for writing the Head block chunks to the disk
+// and access chunks via an mmapped file.
+type HeadReadWriter struct {
```
Can we name it in more explicit way? It's actually very specific head for chunks. So maybe just `HeadChunks`?

codesome replied:
SGTM

codesome replied:
But I see this can be used by anything and not just Head. How about `LiveChunkMmapper`? :P Or `LiveChunkReadWriter`

codesome replied:
Making it `LiveChunkDiskMapper`

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,711 @@
+}
+
+func (w *HeadReadWriter) flushBuffer() error {
+	// Assumes that wbufLock is locked before calling this method.
```
I guess it's not always possible. Also let's add this comment on top of this method.

codesome replied:
> no way try to refactor and have the lock in the funcs where needed?

This flushing is called during operations which are already holding the lock. Unlocking there and locking here is not possible as the critical section is already in progress. I will move the comment on top.

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,711 @@
+func (w *HeadReadWriter) flushBuffer() error {
+	// Assumes that wbufLock is locked before calling this method.
+	if err := w.wbuf.Flush(); err != nil {
+		return nil
```
:+1:

codesome replied:
Yes, will fix, thanks!

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,711 @@
+func (w *HeadReadWriter) shouldCutSegment(chunkLength int, maxt int64) bool {
+	return w.n == 0 || // First segment
+		(maxt-w.curFileMint > w.segmentTime && w.n > HeadSegmentHeaderSize) || // Time duration reached for the existing file.
+		w.n+int64(chunkLength+27) >= MaxSegmentSize
```
Maybe it's @codesome birthday day 27th?  =D

codesome replied:
It should actually be 30 which is the max chunk header size. I will change it to constants.

Spoiler alert: my birthday is in the magic number of these segment files :)

bwplotka commented on tsdb/docs/format/head_chunks.md:
```diff
@@ -0,0 +1,41 @@
+│    mint <8 byte, uint64>     │
+├──────────────────────────────┤
+│    maxt <8 byte, uint64>     │
+├──────────────────────────────┤
```
Do we need crc anywhere here?

codesome replied:
Maybe we do, for mint and maxt. But we can again face a race here if Prometheus crashed while writing crc when maxt was already written (note: maxt is written when closing the segment)

bwplotka replied:
:+1:

bwplotka commented on tsdb/fileutil/mmap.go:
```diff
@@ -28,11 +28,16 @@ func OpenMmapFile(path string) (*MmapFile, error) {
 	if err != nil {
 		return nil, errors.Wrap(err, "try lock file")
 	}
+	defer func() {
```
:clap:

codesome commented:
Thanks @krasi-georgiev and @bwplotka for the reviews. I will start addressing them 1 by 1 now.

codesome commented:
@krasi-georgiev @bwplotka I have addressed most of your comments except those which might require more discussion. This is ready for review again otherwise.

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,748 @@
+			if err != nil {
+				return err
+			}
+			defer f.Close()
```
It would be nice to catch this error with something like https://github.com/thanos-io/thanos/blob/master/pkg/runutil/runutil.go#L133

codesome commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,748 @@
+	return fmt.Sprintf("corruption in segment %s: %s", segmentFile(e.Dir, e.Segment), e.Err.Error())
+}
+
+// LiveChunkDiskMapper is for writing the Head block chunks to the disk
```
Actually the chunk is not live (not being appended to), so `ChunkDiskMapper` sounds fine

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,752 @@
+// and access chunks via an mmapped file.
+type ChunkDiskMapper struct {
+	// Writer.
+	dirFile *os.File
```
file that is directory? (: I would name it more explicit I think...

bwplotka replied:
Maybe just `dir`?

codesome replied:
Yeah it's a directory, I will change the name. As it was os.File, I ended up adding the File suffix :P

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,752 @@
+	curFileMaxt     int64 // In milliseconds.
+	curFileSequence int
+	n               int64 // Bytes written in current open file.
+	maxFileDuration int64 // Duration in milliseconds.
```
Let's remove the comment and add this to var name? `maxFileMs`

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,752 @@
+	n               int64 // Bytes written in current open file.
+	maxFileDuration int64 // Duration in milliseconds.
+
+	buf                 [8]byte
```
Looks like some special buf if 8 bytes.. let's name it `chkBuf`, maybe?

codesome replied:
`writeBuf`, will change to that. It is used while writing a chunk.

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,752 @@
+		return nil, err
+	}
+
+	hrw := &ChunkDiskMapper{
```
```suggestion
	m := &ChunkDiskMapper{
```

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,752 @@
+	return hrw, hrw.openMMapFiles()
+}
+
+func (w *ChunkDiskMapper) openMMapFiles() (returnErr error) {
```
`w` is not right here (: 

@codesome details matter :smile:

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,752 @@
+	wbuf                *bufio.Writer
+	wbufLock            sync.RWMutex
+	crc32               hash.Hash
+	inBufferChunks      [inBufferShards]map[uint64]chunkenc.Chunk
```
Can we comment, where this idea came from? Of sharding this? Why we need this? I don't mind, but would be nice for the reader to know in a comment here.

codesome replied:
The basic requirement was to just store chunk in the buffer till we have enough to flush. As it's a write and read buffer, it needs to be lock protected. If we are writing 1 chunk, all the queries will get blocked which are trying to read from the buffer. Hence the sharding to reduce contention.

I will comment on it.

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,752 @@
+
+	// Reader.
+	// The int key in the map is the file number on the disk.
+	mmappedChunks    map[int]ByteSlice
```
Those are not chunks.. but chunkFiles? 
```suggestion
    mmappedChunksFiles    map[int]ByteSlice
```

bwplotka replied:
Those are not chunks.. but chunkFiles? 
```suggestion
    mmappedChunksFiles    map[int]ByteSlice
```

codesome commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,752 @@
+}
+
+func (e *CorruptionErr) Error() string {
+	return fmt.Sprintf("corruption in head chunk file %s: %s", segmentFile(e.Dir, e.FileIndex), e.Err.Error())
```
When I think of string, Sprintf comes to mind. For this error, Wrapf should be fine.

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -3,7 +3,7 @@
 // You may obtain a copy of the License at
 //
-// http://www.apache.org/licenses/LICENSE-2.0
+// http://wwcdm.apache.org/licenses/LICENSE-2.0
```
bug?

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,753 @@
+}
+
+// flushBuffer flushes the current in-memory chunks.
+// Assumes that wbufLock is _write_ locked before calling this method.
```
This is what I said it's tricky with multi locks ;p

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,753 @@
+	}
+	for i := 0; i < inBufferShards; i++ {
+		cdm.inBufferChunksLocks[i].Lock()
+		cdm.inBufferChunks[i] = make(map[uint64]chunkenc.Chunk)
```
Another reason for using slices instead of maps. We can just reuse underlying array of slice.

bwplotka replied:
And yea, if we can avoid those locks... ^^

codesome replied:
The chunk references here are not sequential. They are randomly spread 64-bit numbers. So handling it via slice is complex and maybe inefficient to search.

codesome replied:
We will need locks for those shards similar to stripeSeries in Head :) The map is not thread-safe.

bwplotka replied:
No if we would lock all

bwplotka replied:
Chunk refs are sequential... if you split the ref by chunkFile ID and chunk ID, no? But maybe I misundertstand the use of this buffer

codesome replied:
As we are flushing the chunks to disk in batches, we cannot mmap it yet before writing to disk. So we store the chunks in this buffer to be served when requested until then.

We have an option of using a single map/slice for this buffer, but to keep it read/write safe we need lock of some kind, and having a single lock would increase the contention in high QPS environments. So I have divided the buffer into shards. And yes, the chunk refs are sequential.

codesome replied:
I moved the entire `inBufferChunks` and its logic to a separate struct called `chunkBuffer` (at the end of the file). I hope now the lock usage feels slightly less complex than before.

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,753 @@
+	return cdm.deleteFiles(segs)
+}
+
+// Size returns the size of the chunks.
```
This comment is lying I think - we return here size of all chunkFiles not chunks itself, no?

bwplotka replied:
Why we need Size?

codesome replied:
I will update the comment. We need this for the size-based retention of TSDB. As these are writing on disk, we need to take this into account too.

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,753 @@
+	oldSeq := cdm.curFileSequence
+	cdm.curFileSequence = seq
+
+	// Write the provided time for mint and 0 for maxt.
```
We can read that from code IMO

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,753 @@
+		cdm.mmappedChunksMtx.Unlock()
+		// Closing the last mmapped file.
+		if err := oldMmapFile.Close(); err != nil {
+			cdm.mmappedChunksMtx.Unlock()
```
double unlock? (Again,... let's reduce locks number)

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,753 @@
+		}
+	}
+
+	mmapFile, err := fileutil.OpenMmapFileWithSize(f.Name(), int(MaxHeadChunkFileSize))
```
If we want those locks (probably not so many), we might want to have some helper for this. (:

codesome replied:
I did not understand what you meant here :)

bwplotka replied:
I mean that we have tons of locks and proposing a helper, but I think you reduce a bit, so will review.

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,753 @@
+		if err := cdm.cut(mint); err != nil {
+			return 0, err
+		}
+		cdm.curFileMaxt = 0
```
It's weird that we set currFileMaxt here, but currFileMin in `cut`. Let's maybe be consistent and choose one place? ( :

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,753 @@
+	// if len(chk.Bytes()) >= writeBufferSize-30, it means that chunk >= the buffer size;
+	// so no need to flush here, as we have to flush at the end (to not keep partial chunks in buffer).
+	if len(chk.Bytes()) < writeBufferSize-30 && cdm.wbuf.Available() < 8+8+8+1+5+len(chk.Bytes()) {
+		if err := cdm.flushBuffer(); err != nil {
```
Can we always just flush?

codesome replied:
That was what I had initially, flush at the end of WriteChunk. But to avoid write amplification, we are buffering it.

krasi-georgiev replied:
hm that is quite interesteding, have you done any research? could there be some way to actually test the difference in the write amplification?

codesome replied:
> Isn't the kernel here keeping some write buffer or is this synced at a set interval

It does keep some buffer but I am not sure how it interacts with flush here. I _guess_ it has to sync with the disk for it to me m-mapped.

And regarding how much amplification we can expect, here is some quick calculation:

SSD page size starts at 2KB (from a quick Google search). And the chunks being written to the disk are under 200-250B (including the meta). Let's consider it to be 256B for easy calculation. So in that case, if we had to flush chunk for every write, we would be doing >8x more disk writes in terms of volume. And this is just for an empty page.

Given a single page can accommodate multiple chunks, overwriting the same page is a bigger task which includes copying+erasing+writing the new updated content at _block_ level. SSD block sizes vary between 256KB-4MB (from a quick Google search). So, in this case, it turns out to be 1204-16384x more writes by volume.

codesome replied:
I think that is one of the reasons we have the "page" concept in the WAL too. And have 8MB write buffer when writing a block. Now that makes me feel I should increase the write buffer size to 4MB here from 2MB that I have right now.

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,753 @@
+		return 0, err
+	}
+
+	n := binary.PutUvarint(cdm.writeBuf[:], uint64(len(chk.Bytes())))
```
Are you not overriding encoding here?

codesome replied:
I did not understand what you mean here

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,753 @@
+		cdm.curFileMint = mint
+		// As this wont happen a whole lot of time, we don't wait
+		// for a new head chunk file to be cut to write it off.
+		binary.BigEndian.PutUint64(cdm.writeBuf[:], uint64(mint))
```
Why it matters? Can we just stick to `cut` and have all in single place for simplicity?

codesome replied:
Maxt depends on the chunks. We cannot have the maxt of all the chunks to be written to the file in future in `cut` :). So we update the maxt whenever we write chunks.

codesome replied:
We need maxt for the truncation of files. (Although you could argue that you can get away with mint)

codesome replied:
Turns out I gave a wrong explanation here. So we write the maxt only at the end, and if Prometheus happens to crash, maxt remains 0 in the file. So we are maintaining mint in on the disk, which is checked at every WriteChunk. Using this mint, we can optimistically set the maxt for the crashed file when we are getting Prometheus back up (to help with the truncation).

codesome replied:
Another thing we could do is, when we get a new chunk to write after Prometheus comes back, we can use it's mint as maxt for all the old files which have 0 as maxt (which means crash). This could mean a delayed deletion of files (which might hurt the actual data retention?).

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,753 @@
+	cdm.inBufferChunksLocks[chunkRef%inBufferShards].Unlock()
+
+	if len(chk.Bytes()) >= writeBufferSize-30 {
+		if err := cdm.flushBuffer(); err != nil {
```
Hm, it will be done on the begining of WriteChunk?  also why not flushing always (again)

codesome replied:
I am not sure why I have this flush here. I will remove it as it seems unnecessary.

codesome replied:
Ah I see now why. If the chunk size was more than the max buffer size itself, then there is no point keeping partial data in the buffer, so flush everything at the end. I will add a comment there, I think this check should be there.

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,753 @@
+}
+
+// Chunk returns a chunk from a given reference.
+func (cdm *ChunkDiskMapper) Chunk(ref uint64) (chunkenc.Chunk, error) {
```
Note that the returned chunk's bytes are not available after Close of ChunkDiskMapper maybe?

codesome replied:
Will add that in the comment.

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,753 @@
+	cdm.mmappedChunksMtx.RLock()
+	// We hold this read lock for the entire duration because if the Close()
+	// is called, the data in the byte slice will get corrupted as the mmapped
+	// file will be closed.
```
I think that's kind of obvious, I would skip this comment.

codesome replied:
I would like to keep the comment. Initially, I did not have the lock for the entire code and spent a lot of time debugging to find out why am I getting corrupt data :). So I added the comment after I realized it to not repeat the same mistake in that place.

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,753 @@
+	// file will be closed.
+	defer cdm.mmappedChunksMtx.RUnlock()
+
+	if cdm.closed {
```
Do we needmapped chunks lock for this? 

Again... too many locks ;p

codesome replied:
Yeah, we kind of need. Else there would be a race between `Close()` call and this for updating the `closed` variable.

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,753 @@
+
+// IterateAllChunks iterates on all the chunks in its byte slices in the order of the head chunk file sequence
+// and runs the provided function on each chunk. It returns on the first error encountered.
+func (cdm *ChunkDiskMapper) IterateAllChunks(f func(seriesRef, chunkRef uint64, mint, maxt int64) error) error {
```
Is this really needed? It is used only by tests? I guess, yes for DB cut? If no let's move it to test maybe? Or unexport?

codesome replied:
It will be used by Head when we will replay chunks from the disk. It is already integrated in the parent PR, but there is no use of this as a standalone function in this PR.

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,753 @@
+	chkCRC32 := newCRC32()
+
+	// Iterate files in ascending order.
+	seqs := make([]int, 0, len(cdm.mmappedChunkFiles))
```
We would not need that if it was slice as I propose.

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,753 @@
+			}
+			chkCRC32.Reset()
+			startIdx := idx
+			seriesRef := binary.BigEndian.Uint64(bs.Range(idx, idx+8))
```
you already have nice consistants for this.

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,753 @@
+			maxt := int64(binary.BigEndian.Uint64(bs.Range(idx, idx+8)))
+			idx += 8
+
+			chunkRef := uint64(seq)<<32 | uint64(idx)
```
We already have nice helper for this I think

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,753 @@
+	return chunkRef, nil
+}
+
+func (cdm *ChunkDiskMapper) chunkRef(seq, offset uint64) (chunkRef uint64) {
```
Does not need to be method, it can be just function.

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,753 @@
+			idx += 8
+
+			chunkRef := uint64(seq)<<32 | uint64(idx)
+			if err := f(seriesRef, chunkRef, mint, maxt); err != nil {
```
Wait, why we cannot use `Chunk(..)` method for all of this?

codesome replied:
1. We don't have the ref yet to call `Chunk(...)`
2. We don't need to chunk data here actually, whereas `Chunk(...)` returns the actual chunks. What we need here are the ones passed as an argument, which is not given by `Chunk(...)`.

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,753 @@
+	var removedFiles []int
+
+	cdm.mmappedChunksMtx.RLock()
+	for seq, bs := range cdm.mmappedChunkFiles {
```
When we have it sorted we can just abort when we see greater mint

codesome replied:
It's a map, so it won't be sorted always I guess :). Regarding using a slice instead, I have comment above it above.

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,753 @@
+		closers = append(closers, cdm.closers[seq])
+		delete(cdm.mmappedChunkFiles, seq)
+		delete(cdm.closers, seq)
+		if err := os.Remove(segmentFile(cdm.dir.Name(), seq)); err != nil {
```
I think that's why windows tests fails. You cannot delete before closing.

codesome replied:
Windows tests failing here are because the machine is not able to handle the mmap size (which hides this one I guess). Good point! I will fix it

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,753 @@
+func (cdm *ChunkDiskMapper) deleteFiles(removedFiles []int) error {
+	closers := make([]io.Closer, 0, len(removedFiles))
+	cdm.mmappedChunksMtx.Lock()
+	for _, seq := range removedFiles {
```
Let's add some newlines, it's huge block of text

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,753 @@
+	// Delete all the head chunk files following the corrupt head chunk file.
+	segs := []int{}
+	for seg := range cdm.mmappedChunkFiles {
+		if seg >= cerr.FileIndex {
```
This will straight away with slice I think

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,753 @@
+// (including the corrupt file).
+func (cdm *ChunkDiskMapper) Repair(originalErr error) error {
+	err := errors.Cause(originalErr) // So that we can pick up errors even if wrapped.
+	cerr, ok := err.(*CorruptionErr)
```
Do we need expose publically `CorruptionErr`?

codesome replied:
Maybe not. I will make it internal for now.

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,753 @@
+		if err != nil {
+			return errors.Wrap(err, "mmap files")
+		}
+		cs[seq] = f
```
Slices might be better because:
* You don't need to rewrite + sort those maps all the time
* Sometimes because of the above, you just not sort, and do more inefficient stuff (e.g search)
* With slice you can then reuse underlying array when you reuse (e.g in bufferedChunks)

codesome replied:
> We can operate on just slice indexes, right? Will be much cleaner and deterministic, no?

The sequence does not always start at 0. After a point once we start deleting old files, it will never start from 0, and keeping track of offset and using a slice would be more buggy and complex. Map feels simpler here. 

Plus, there won't be a lot of files, so I would not worry about re-use and search here, maps would be quick enough. And we only sort the files in `IterateAllChunks` which will be done only once on start, so I would not worry about that too.

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,753 @@
+
+	writeBuf [8]byte
+	wbuf     *bufio.Writer
+	wbufLock sync.RWMutex
```
I am worried essentially that we sacrifice complexity for micro-optimizations.

codesome replied:
I looked at the locks, and I think 2/3 locks would make sense here for high load. One lock dedicated to read and write each (more on the in-buffer chunk lock at the end, we can get rid of this).

Now talking about `wbufLock` (the lock for the writes) and `mmappedChunksMtx` (the lock for reads), having a common lock has the problem of locking the hot-path of either queries or ingestion where there would be potentially millions of chunks being written and hundreds of queries simultaneously. So it doesn't look like a micro-optimization to have 2 locks here 

> usually when you modify mmapedChunkFiles, you modify wbuf as well

Yeah, but that happens only when the file is full, so 1 write with both locks held (when creating a new file) vs 100s of writes with only 1 lock held when writing to same file - I guess for majority case separate lock makes sense to reduce contention.

> Also not sure if I get why you have locks per shard if you lock wfbuf anyway.

You are right, I think we can get of it with what you said. I will verify once again and remove it.

codesome replied:
Looking at the code once again, I think we can reduce the contention between read and write more by not using the `wbufLock` in the read path and instead only make use of the in-buffer lock by skipping a check. I will explore this option.

codesome commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,753 @@
+
+func (cdm *ChunkDiskMapper) shouldCutNewFile(chunkLength int, maxt int64) bool {
+	return cdm.n == 0 || // First head chunk file.
+		(maxt-cdm.curFileMint > cdm.maxFileMs && cdm.n > HeadChunkFileHeaderSize) || // Time duration reached for the existing file.
```
Time: so that we can delete the files kind of deterministically after some time. File deletion is based on time.
Size: OS has limits on how big of a file we can mem-map, so we limit at that too.

bwplotka replied:
Hm, can we mention that in comment. I think it's unlikely we get out OS limit, but sure (:

codesome replied:
`MaxInt32` (the max size for file now) it just ~1GiB :) so it is easily possible to cross it in high load environments. I will add a comment, and maybe also explicitly have it as 1 GiB instead of MaxInt32.

codesome commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,753 @@
+		}
+	}()
+
+	files, err := sequenceFilesMap(cdm.dir.Name())
```
I am using the map below, so I guess we will have the keep the map around. I will change the function name.

codesome commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,753 @@
+			return errors.Wrap(err, "mmap files")
+		}
+		cs[seq] = f
+		bs[seq] = realByteSlice(f.Bytes())
```
https://github.com/prometheus/prometheus/pull/6830#discussion_r385533675

codesome commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,753 @@
+		seqs = append(seqs, seq)
+	}
+
+	cdm.mmappedChunkFiles = bs
```
https://github.com/prometheus/prometheus/pull/6830#discussion_r385533675

codesome commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,753 @@
+	cdm.closers = cs
+	cdm.size = 0
+
+	// Check for unsequential files.
```
Yeah, gaps, will update the comments

codesome commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,753 @@
+	}
+
+	for i, b := range cdm.mmappedChunkFiles {
+		if b.Len() < HeadChunkFileHeaderSize {
```
It will be automatic once we integrate this with Head. It works similar to how WAL repair works now. WAL replay fails in Head, and then we attempt the repair after that. I have it in place in the patent PR where after failing to replay the mmapped chunks, we attempt the repair.

codesome replied:
It turns out there is repair in place for errors on `IterateAllChunks`. I will update the parent PR after this PR is merged to repair errors from here too.

codesome commented:
I tried having 2 different structs for read and write path as @krasi-georgiev suggested. Though individual structs are simpler within themselves, using them together causes additional complexity (like a couple of shared variables to be duplicated, interaction during cutting a new file, etc). So I decided to go with a single struct which is simpler in this case.

krasi-georgiev commented on tsdb/chunks/head_chunks_test.go:
```diff
@@ -0,0 +1,301 @@
+	testutil.Equals(t, expectedBytes, actualBytes[HeadChunkFileHeaderSize:])
+}
+
+func TestHeadReadWriter_ReadChunk(t *testing.T) {
```
I still think that you can combine this test with the previouse one, the only dif I see that you just check the actual chunk content so this should just add no more than few lines to the test above, no?

codesome replied:
Another difference is that in WriteChunk I am only filling up 1 file and checking it's contents. In ReadChunk, I am having >1 files. So while combining those tests is possible, the simplicity of reading the test will go away by trying to handle that case.

And I feel explicitly splitting the reading and writing tests would make it easier to find bugs if any. Open for opinions on this.

krasi-georgiev replied:
yep, I don't mind, maybe if you put the boostraping for the data in a separate func and reuse it in all tests?

krasi-georgiev replied:
ok I just had another look and yes one test is named read and the other is named write, but both write the chunk to disk one both read it after that and theck different bits, so unless combining them would make it very very hard to follow I would vote to have it as a single one.

krasi-georgiev commented on tsdb/chunks/head_chunks_test.go:
```diff
@@ -0,0 +1,301 @@
+	}
+}
+
+func TestHeadReadWriter_IterateChunks(t *testing.T) {
```
the same here you jast add an additional check for the iterator. the boostrap is quite similiar.

If you prefer to keep those separate can't you make the data boostrapper in a seprate func to be used in all tests?

codesome replied:
Yeah, so this can be combined with `ReadChunks` test and would work out well too as in the read test we are assuring >1 files. I will combine these 2.

krasi-georgiev commented on tsdb/chunks/head_chunks_test.go:
```diff
@@ -0,0 +1,301 @@
+		_, err := hrw.WriteChunk(1, int64((i-1)*100)+1, int64(i*100)-1, randomChunk(t))
+		testutil.Ok(t, err)
+
+		time.Sleep(100 * time.Millisecond)
```
why is this sleep here?

codesome replied:
I have no idea! Will remove it

krasi-georgiev commented on tsdb/chunks/head_chunks_test.go:
```diff
@@ -0,0 +1,301 @@
+		testutil.Ok(t, hrw.cut(int64(i-1)*100))
+
+		// Write a chunks to set maxt for the segment.
+		_, err := hrw.WriteChunk(1, int64((i-1)*100)+1, int64(i*100)-1, randomChunk(t))
```
hm , are you sure this can't be simplified a bit? maybe it is just me, but I don't want to do match when reading tests :)

codesome replied:
It can be simplified by not using `i` and just incrementing some variable :P will make the change

codesome replied:
Using the `i` felt intuitive to be TBH at you can guess the ranges of the previous and the next file

krasi-georgiev replied:
much better, thanks.

krasi-georgiev commented on tsdb/chunks/chunks.go:
```diff
@@ -206,7 +206,7 @@ func (w *Writer) cut() error {
 }
 
-func cutSegmentFile(dirFile *os.File, chunksFormat byte, segmentSize int64) (headerSize int, newFile *os.File, seq int, err error) {
+func cutSegmentFile(dirFile *os.File, magicNumber uint32, chunksFormat byte, segmentSize int64) (headerSize int, newFile *os.File, seq int, err error) {
```
maybe ?
```suggestion
func cutSegmentFile(dirFile *os.File, magicNumber uint32, chunksFormat byte, allocSize int64) (headerSize int, newFile *os.File, seq int, err error) {
```

krasi-georgiev commented on tsdb/chunks/chunks.go:
```diff
@@ -226,7 +226,7 @@ func cutSegmentFile(dirFile *os.File, chunksFormat byte, segmentSize int64) (hea
 	// Write header metadata for new file.
 	metab := make([]byte, SegmentHeaderSize)
-	binary.BigEndian.PutUint32(metab[:MagicChunksSize], MagicChunks)
+	binary.BigEndian.PutUint32(metab[:MagicChunksSize], magicNumber)
```
yes but why does it have to be like a func argument, the chunk `cutSegmentFile` will allways cut chunks to that magic number would always be `MagicChunks` so why not use it directly inside the func

codesome replied:
`cutSegmentFile` is used for both block segments and the head chunks file that this PR introduces.

For head chunk files, we use a different magic number, see: https://github.com/prometheus/prometheus/blob/e289698d97d3b21d97c921c3729496a1157f80ea/tsdb/chunks/head_chunks.go#L365

codesome commented on tsdb/chunks/head_chunks_test.go:
```diff
@@ -0,0 +1,301 @@
+
+	expectedBytes := []byte{}
+	numChunks := 2000
+	nextChunkOffset := uint64(HeadChunkFileHeaderSize + SeriesRefSize + 2*MintMaxtSize)
```
So I kept the offset for chunk to be the place where the actual chunk data starts (i.e. the encoding, data length, and the data itself), because, with the chunk reference (the offset), we never try to read the series reference or the mint maxt. But if that looks confusing, I can make the offset to be at the series reference.

krasi-georgiev replied:
yes please unless it makes something else more complicated.

codesome replied:
Done

codesome commented on tsdb/chunks/head_chunks_test.go:
```diff
@@ -0,0 +1,301 @@
+		// += encoding + chunk data len + chunk data + CRC + seriesRef,mint,maxt of next chunk.
+		nextChunkOffset += 1 + uint64(n) + uint64(len(chunk.Bytes())) + CRCSize + SeriesRefSize + 2*MintMaxtSize
+	}
+	testutil.Ok(t, hrw.flushBuffer())
```
Will use a `Close()` here

codesome commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,782 @@
+		return 0, err
+	}
+
+	if err := cdm.write(cdm.crc32.Sum(cdm.writeBuf[:0])); err != nil {
```
`writeAndAppendCRC` and `writeCRC` sounds good. We wont need `resetCRC` explicitly as there is `crc32.Reset()` already.

krasi-georgiev commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,786 @@
+	return cdm.size + n
+}
+
+func (cdm *ChunkDiskMapper) Close() error {
```
comment for exported method

krasi-georgiev commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,786 @@
+	chkRef = chunkRef(uint64(cdm.curFileSequence), uint64(cdm.n))
+
+	binary.BigEndian.PutUint64(cdm.writeBuf[:], seriesRef)
+	if err := cdm.writeAndAppendToCRC32(cdm.writeBuf[:SeriesRefSize]); err != nil {
```
why not make the buf big enough to hold all the chunk metadata, and than you can write to the buffer only 2 times 1 for the metadata and one for the chunks data? I mean why do you have to write to buffer for every chunk element?

codesome replied:
Sounds fair and that will clean-up the code too. Max meta size is not big enough to be a concern. I will change it.

krasi-georgiev commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,786 @@
+
+	if mint < cdm.curFileMint {
+		cdm.curFileMint = mint
+		// As this wont happen a whole lot of time, we don't wait
```
```suggestion
		// As this won't happen a whole lot of time, we don't wait
```

krasi-georgiev commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,786 @@
+}
+
+// WriteChunk writes the chunk to the disk.
+// The returned chunk ref is the reference from where the chunk encoding starts for the chunk.
```
Without having a context it is unclear why we need maxt and mint here. I think it would be usefull to either explain here or in the head chunk readme or anywhere else.

codesome replied:
I will add it in the docs where we have the layout of the chunks written.

codesome commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,786 @@
+		if err != nil {
+			return errors.Wrap(err, "mmap files")
+		}
+		cs[seq] = f
```
For this part of code, this looks fine

codesome commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,786 @@
+	lastSeq := chkFileIndices[0]
+	for _, seq := range chkFileIndices[1:] {
+		if seq != lastSeq+1 {
+			return errors.Errorf("found unsequential head chunk files %d and %d", lastSeq, seq)
```
`chkFileIndices` is sorted here, so no problem here. But good point, we should sort it even during truncate. I will make that change.

krasi-georgiev replied:
Yes this is what I meant these need to be sorted when deleting.

codesome commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,786 @@
+		maxt := binary.BigEndian.Uint64(b.Range(HeaderMaxtOffset, HeaderMaxtOffset+MintMaxtSize))
+		if maxt == 0 {
+			// This is possible if Prometheus crashes and the maxt was unwritten to the
+			// last head chunk file. As a safe buffer, we set the maxt to mint + 1.5 times head chunk file time range.
```
> Can you remind me how is this maxt used?

maxt is used during truncation. If the file's maxt is less than the time passed in the Truncate method, then it is deleted.

> Is this really safe to do?

I am still trying to evaluate this myself. Given during writing chunks we ensure that chunks don't cross the max file time range, I thought having it 1.5* the max file time rang should be safe enough.

codesome commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,786 @@
+func (cdm *ChunkDiskMapper) shouldCutNewFile(chunkSize int, maxt int64) bool {
+	return cdm.n == 0 || // First head chunk file.
+		(maxt-cdm.curFileMint > cdm.maxFileMs && cdm.n > HeadChunkFileHeaderSize) || // Time duration reached for the existing file.
+		cdm.n+int64(chunkSize+MaxHeadChunkMetaSize) >= MaxHeadChunkFileSize // Exceeds the max head chunk file size.
```
`>` should do it.

codesome commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,786 @@
+	curFile *os.File // File being written to.
+
+	curFileMint     int64 // In milliseconds.
+	curFileMaxt     int64 // In milliseconds.
```
I will add more comments

krasi-georgiev commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,786 @@
+
+func (cdm *ChunkDiskMapper) openMMapFiles() (returnErr error) {
+	bs := map[int]ByteSlice{}
+	cs := map[int]io.Closer{}
```
Did you forget to comment or push the changes? This and few other places were marked as resolved without any comment or code changes/

codesome replied:
I tend to mark as resolve as and when I am coding it locally so that it is easier to keep track of unresolved comments (when we have so many comments). I will be pushing the updates shortly.

codesome commented on tsdb/docs/format/head_chunks.md:
```diff
@@ -0,0 +1,43 @@
+
+# Chunk
+
+Unlike chunks in the on-disk blocks, here we additionally store series reference that the chunks belongs to and the mint/maxt of the chunks. This is because we don't have an index associated with these chunks, hence these meta information are used while replaying the chunks.
```
Yes, mint maxt is obtained from chunks itself. And right, we would be replaying all the chunks on start, so we should be able to calculate mint-maxt from there (or maybe just maxt is enough after that). I will add it in the comments stating that `IterateAllChunks` should be called after initializing the struct to set those properly (there is where the iteration happens)

codesome replied:
I think I misunderstood your comment. If you meant getting mint and maxt of chunks from the data itself, it would not scale. We would need to iterate each and every sample to do that. When we want to restore the chunks in the memory, we need to know the mint and maxt of the chunks so that we can perform queries over them, hence we need them here.
PS: Even for persistent blocks, we store the mint and maxt separately in the index.

If you meant the above, then with this misunderstanding I actually got rid of mint and maxt for the file. As we would iterate the chunks once anyway on startup for the replay, I am restoring  the mint maxt of the files during that time. This would mean calling `IterateChunks` at least once (which we will do).

krasi-georgiev replied:
yep that was my point, thanks , that simplifies the implementation a bit.

krasi-georgiev replied:
> If you meant the above, then with this misunderstanding I actually got rid of mint and maxt for the file. As we would iterate the chunks once anyway on startup for the replay, I am restoring the mint maxt of the files during that time. This would mean calling IterateChunks at least once (which we will do).

I am confused, this comments reads that you have removed mint and maxt from the file, but it is still there in the code?

I am missunderstanding something

codesome replied:
https://github.com/prometheus/prometheus/blob/012161d90d6a8a6bb930b90601fb89ff6cc3ae60/tsdb/head.go#L2017-L2020

So a chunk does not have mint and maxt with it explicitly, it's just in the samples. So, to fill in mint and maxt in `memChunk` or the future mmaped chunk, we need this mint and maxt. 

Similarly for blocks, the mint and maxt of chunks are stored for all chunks in the index https://github.com/prometheus/prometheus/blob/012161d90d6a8a6bb930b90601fb89ff6cc3ae60/tsdb/docs/format/index.md#series

krasi-georgiev replied:
so I guess that is used for quering mainly? So you know whih chunks to include when requesting some data?

codesome replied:
Yes. (And in case of mmaping, also to know max time of the file for truncation)

codesome commented:
Now it's time to fix those windows test finally! I tried reducing the max file size to 128 MiB, and still it fails with `mmap: CreateFileMapping: Not enough memory resources are available to process this command.`.

krasi-georgiev commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,767 @@
+
+	// Iterate files in ascending order.
+	segIDs := make([]int, 0, len(cdm.mmappedChunkFiles))
+	for seg := range cdm.mmappedChunkFiles {
```
are you sure we don't have a race here for the mmappedChunkFiles map with the WriteChunk method?

codesome replied:
So, `IterateAllChunks` is not supposed to be called in parallel with `WriteChunk` or `Truncate` methods (and we won't be doing it too, only called in the beginning to replay). But just as a safety net, I will hold the write path lock here.

krasi-georgiev commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,767 @@
+		mmapFile := cdm.mmappedChunkFiles[segID]
+		sliceLen := mmapFile.byteSlice.Len()
+		idx := HeadChunkFileHeaderSize
+		for idx < sliceLen {
```
how about 
```suggestion
		fileEnd := mmapFile.byteSlice.Len()
		idx := HeadChunkFileHeaderSize
		for idx < fileEnd {
```

codesome commented on tsdb/chunks/head_chunks_test.go:
```diff
@@ -0,0 +1,278 @@
+
+			binary.BigEndian.PutUint64(buf[:], uint64(mint))
+			expectedBytes = append(expectedBytes, buf[:MintMaxtSize]...)
+			_, err = chkCRC32.Write(buf[:MintMaxtSize])
```
I can increase the buffer size and do it. Will do it.

codesome commented on tsdb/chunks/head_chunks_test.go:
```diff
@@ -0,0 +1,278 @@
+		}
+	}
+
+	/// Checking on-disk bytes for the first file.
```
I used `///` to mark the start of the new section of testing as I had merged multiple tests. Maybe not required actually.

codesome commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,767 @@
+	curFileMint     int64    // Current file's mint. Used to check for a chunk crossing the max file time range.
+	curFileMaxt     int64    // Maxt of all chunks written to current file. This is written to the file when the file is closed.
+	curFileNumBytes int64    // Bytes written in current open file.
+	maxFileMs       int64    // Max time range (curFileMaxt-curFileMint) for a file.
```
Yes, this is better. `Ms` was a leftover from a previous version.

codesome commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,767 @@
+	cdm.curFileSequence = seq
+	cdm.curFileMint = mint
+	cdm.curFile = f
+	if cdm.wbuf != nil {
```
It will require a new file to be cut in `newChunkDiskMapper` for this to be set, which I don't suggest (as it can lead to many empty files). A similar logic was used in writing blocks to avoid having any empty files.

krasi-georgiev replied:
Yeo I see what you mean

codesome commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,767 @@
+
+	if cdm.curFile != nil {
+		// Open it again with the new size.
+		newTailFile, err := fileutil.OpenMmapFile(cdm.curFile.Name())
```
So, the finalized file is already mmapped with an optimistic big size. Here we actually m-map it again with the actual size of the file and close the old m-mapped file. I will update the comment to be more clear.

codesome commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,767 @@
+		}
+		cdm.readPathMtx.Unlock()
+		// Closing the last mmapped file.
+		if err := oldFileCloser.Close(); err != nil {
```
I was guessing this would be confusing :)

So, `oldFileCloser` and `cdm.curFile` are 2 different things.
1. The closers have closers to the m-mapped files which have a read-only file open within them. Hence `oldFileCloser` is for the m-mapped file.
2. `cdm.curFile` is the only write-only file open, and, it is closed in `finalizeCurFile`.

This means the last file is open in 2 places - m-mapping and writing. So I guess no changes here and the comments should be there I guess to explicitly say its for the m-mapped file.

krasi-georgiev replied:
but in this case it is the same file no?

codesome replied:
Same file, but opened in 2 places, so we should close both the object. The files objects are not shared as one of them is read-only.

codesome replied:
To other reviewers: After some offline discussion, we came to conclusion that we won't require reloading the mmapped file again and we can reuse the already mmapped file with a larger initial size.

codesome commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,767 @@
+	curFile         *os.File // File being written to.
+	curFileSequence int      // Index of current open file being appended to.
+	curFileMint     int64    // Current file's mint. Used to check for a chunk crossing the max file time range.
+	curFileMaxt     int64    // Maxt of all chunks written to current file. This is written to the file when the file is closed.
```
Not sure about this suggestion, but we need this `curFileMaxt`. 

When new file is cut, this `curFileMaxt` is used to mark the maxt of that file
https://github.com/prometheus/prometheus/blob/6eefadfeaec25d3fb249d2fafe7608541dc00dcd/tsdb/chunks/head_chunks.go#L367-L370

Which is used during truncation
https://github.com/prometheus/prometheus/blob/6eefadfeaec25d3fb249d2fafe7608541dc00dcd/tsdb/chunks/head_chunks.go#L637

codesome replied:
The mint and maxt was only removed from the on-disk file, but we still need them in the memory :)

krasi-georgiev replied:
yep, yep, I missed that part.

codesome commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,767 @@
+}
+
+// Chunk returns a chunk from a given reference.
+// Note: The returned chunk will turn invalid after closing ChunkDiskMapper.
```
We already check for the closed and return an error. What that comments means is that: If you had got a chunks before closing, then that earlier valid chunk would turn invalid after you close `ChunkDiskMapper`.

codesome commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,767 @@
+	maxFileMs       int64    // Max time range (curFileMaxt-curFileMint) for a file.
+
+	writeBuf     [MaxHeadChunkMetaSize]byte
+	wbuf         *bufio.Writer
```
> if they have their own struct with methods

Went through this and I think it might not simplify it much. `curFile` and it's info (`curFileXXX` variables) is also used with the reader variables (like when cutting a new file). Will skip this grouping for now, but will change that variable name as suggested.

krasi-georgiev commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,767 @@
+	curFile         *os.File // File being written to.
+	curFileSequence int      // Index of current open file being appended to.
+	curFileMint     int64    // Current file's mint. Used to check for a chunk crossing the max file time range.
+	curFileMaxt     int64    // Maxt of all chunks written to current file. This is written to the file when the file is closed.
```
The var name is quite clear that it is for the current file maxt so I am thinkig a comment as simple as.
```suggestion
	curFileMaxt     int64    // Used for the size retention.
```

codesome replied:
More like for `time retention`. Updated the comments.

codesome commented:
Thanks for the reviews @krasi-georgiev and @bwplotka! I will merge this once I find the fix for the windows test. Any further changes can be taken care in https://github.com/prometheus/prometheus/pull/6679

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,767 @@
+
+// NewChunkDiskMapper returns a new writer against the given directory
+// using the default head chunk file duration.
+// NOTE: 'IterateAllChunks' method needs to be called at least once after creating ChunkDiskMapper
```
This is weird - just run IterateAll Chunks in constructor? ;p

codesome replied:
Yeah, that would have been good, but, `Head` would not like it :P. The initialization and replay are separated in `Head` (`NewHead` and `Init`), so we need to have it separately here.

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,767 @@
+			startIdx := idx
+			seriesRef := binary.BigEndian.Uint64(mmapFile.byteSlice.Range(idx, idx+SeriesRefSize))
+			idx += SeriesRefSize
+
```
So many unnecessary new lines IMO

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,767 @@
+			}
+
+			idx += CRCSize
+
```
ditto... do we need those newlines? new lines should separate different logic

codesome commented:
So here is what I have concluded about the windows error: If we set the mmap size of the file to be greater than the file size (which is the case when we cut a new file), it errors out with a Windows bug saying `Not enough memory resources are available to process this command`. Preallocating the file would be a bad idea for the fast path, and also poses many challenges as there is no separate index to access the file. 

Any suggestions on what can be done here for m-mapping an empty file?

bwplotka commented:
Only when new file is cut? Let's focus on this path. Can we create fake empty file for X bytes? and mmap that?

codesome commented:
Yes, that would be the preallocation. As said in the previous comment, we want to avoid that as it lies on the hot path. Also, while iterating the chunks for replay we rely on the file end (maybe we can get around that, by exiting the replay when we encounter all 0's for a chunk and its meta). So the main concern would be stalling the scraping.

bwplotka commented:
Can we do that for windows only?

bwplotka commented:
With build tags

codesome commented:
We can, though not a big fan of it :| I don't see any other option.

codesome commented:
Tried some quick fix. If that fails too, I will give it a shot tomorrow.

codesome commented:
Windows test is finally green! If @krasi-georgiev or @bwplotka can review the last commit that would be great.

krasi-georgiev commented:
did you actually test if preallocating a file really takes more time? I am thinking maybe it shouldn't take longer if the file is only zeros.

codesome commented:
> did you actually test if preallocating a file really takes more time? I am thinking maybe it shouldn't take longer if the file is only zeros.

Apparently it doesn't seem to take very long, 1-1.5ms in my system, but I guess at the end if will depend on the hardware and the current situation of the system. Are you suggesting we have preallocation by default? (We could do that now, and if we face any problems, we can revert for the `!windows`)

codesome commented:
Let me actually just have preallocation by default. We will know about its effect when we run it and changing it will be a quick fix.

codesome commented:
I have made preallocation of file a default now

krasi-georgiev commented:
yep, in that case no need to do another review from my side
LGTM

codesome commented:
Further integration will be in https://github.com/prometheus/prometheus/pull/6679
